### PR TITLE
[Epic] EIP-1271 support

### LINKED ISSUE #1088: [Epic] EIP-1271
## What is the feature about

Implementing off-chain EIP-1271 support into the Safe UI.

## Transaction service endpoints

- [x] Get (all)/propose messages https://github.com/safe-global/safe-transaction-service/pull/1138
  - [x] https://github.com/safe-global/safe-transaction-service/pull/1170
  - [x] https://github.com/safe-global/safe-transaction-service/pull/1176
  - [ ] Fix for safes <1.3.0 https://github.com/safe-global/safe-eth-py/issues/421

## Gateway endpoint

- [x] Get all messages https://github.com/safe-global/safe-client-gateway/issues/998
  - [x] https://github.com/safe-global/safe-client-gateway/pull/1007
- [x] Propose message https://github.com/safe-global/safe-client-gateway/pull/1006
- [x] Confirm message https://github.com/safe-global/safe-client-gateway/pull/1008
- [x] Get message by hash https://github.com/safe-global/safe-client-gateway/pull/1014
- [x] Add `messagesTag` to Safe info endpoint https://github.com/safe-global/safe-client-gateway/pull/1013

## Gateway SDK

- [x] Add endpoints https://github.com/safe-global/safe-gateway-typescript-sdk/issues/98
  - [x] Add singular endpoint https://github.com/safe-global/safe-gateway-typescript-sdk/pull/104

## Web

- [x] "Messages" routes/tab #1155
- [x] Accordion summary #1156
- [x] Accordion details #1157
- [x] Accordion signature stepper #1158
- [x] Fetching of messages #1159
  - [x] Store slice #1217
  - [x] Integrate endpoint #1243
  - [x] Poll on tag change #1273 
- [x] Add new event bus #1205
- [x] New signing modal #1206
  - [x] Modal #1224
  - [x] Integrate with `communicator` #1270
  - [x] Integrate endpoint #1243
- [x] Extend the test Safe App #1207
- [x] Create dApp that uses WC https://github.com/5afe/eip-1271-dapp
- [x] Normalize messages before submitting to backend #1517

## Safe Apps

- [x] Set off-chain signing https://github.com/safe-global/safe-apps-sdk/pull/407
- [x] Update WC Safe App https://github.com/safe-global/safe-react-apps/pull/608

## Links

- [Problem statement w/ kickoffs](https://www.notion.so/EIP-1271-signatures-can-only-be-generated-on-chain-2800b5e516124d959c38ba8f2042f9f7)
- [Backend specifications](https://www.notion.so/gnosis-safe/EIP-1271-backend-specifications-c0c6ef187425475288216021a7521c82?d=3345071f763246439acfe3c765645369#ad5426ada7c54158b8b04b9a4ca25bac)
- [Figma](https://www.figma.com/file/x1uuIJsOZkyv5cwNlDdVC1/EIP-1271?node-id=1%3A2)

## Test Cases

- [EIP-1271 test cases](https://www.notion.so/gnosis-safe/39fa9f1ede724bda989381fb41f0a14c?v=20d51e02d9e748948f1d1d59f6803226)

##Found issues

- [x] https://github.com/safe-global/web-core/issues/1435
- [x] https://github.com/safe-global/web-core/issues/1432
- [x] https://github.com/safe-global/web-core/issues/1431
- [x] https://github.com/safe-global/web-core/issues/1430
- [x] https://github.com/safe-global/web-core/issues/1438
- [x] https://github.com/safe-global/web-core/issues/1447

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Resolves #1088

### `web-core`
- [x] "Messages" routes/tab #1155
- [x] Accordion summary #1156
- [x] Accordion details #1157
- [x] Accordion signature stepper #1158
- [x] Fetching of messages #1159
  - [x] Store slice #1217
  - [x] Integrate endpoint #1243
  - [x] Poll on tag change #1273 
- [x] Add new event bus #1205
- [x] New signing modal #1206
  - [x] Modal #1224
  - [x] Integrate with `communicator` #1270
  - [x] Integrate endpoint #1243
- [x] Extend the test Safe App #1207

### Safe Apps
- [x] Set off-chain signing https://github.com/safe-global/safe-apps-sdk/pull/407
- [x] Create [dApp that uses WC](https://github.com/5afe/eip-1271-dapp)
- [x] Update the WC Safe App https://github.com/safe-global/safe-react-apps/pull/608

## Found issues 

- [x] https://github.com/safe-global/web-core/issues/1435
- [x] https://github.com/safe-global/web-core/issues/1432
- [x] https://github.com/safe-global/web-core/issues/1431
- [x] https://github.com/safe-global/web-core/issues/1430
- [x] https://github.com/safe-global/web-core/issues/1438

## Known issues

- Nested Safes do not receive signatures from the parent

## How to test it

Prerequisites:
- Open the https://github.com/safe-global/safe-react-apps/pull/608 deployment.
- Run the [EIP-1271 test DApp](https://github.com/5afe/eip-1271-dapp) locally.

Method:
1. Open the aforementioned WC app deployment.
2. Connect to the aforementioned test DApp.
3. Toggle off-chain signing and attempt to sign a message.
4. Observe the off-chain signing modal.
5. Sign the message.
6. Observe the signed message in the "Signatures" tab of the transaction list.

## Analytics changes

1. Signing a message dispatches a "Sign message"/"Confirm message" "wallet" event accodingly.
2. Clicking "Sign" dispatches a "Sign message" "tx-list" event.
3. `safe_setSettings` RPC calls should be tracked when enabling/disabling off-chain signing.

### CODE REVIEW

File: src/components/safe-messages/MsgModal/index.tsx [hunk 1]
```diff
@@ -111,6 +112,8 @@ const MsgModal = ({
         await dispatchSafeMsgConfirmation({ signer, safe, message: decodedMessage, requestId })
       }
 
+      trackEvent(MESSAGE_MODAL_EVENTS.SIGN_MESSAGE)
```
Comment (katspaugh) [inline src/components/safe-messages/MsgModal/index.tsx:115]: Please don't trigger events like this, there's `useTxTracking` that subscribes to events and tracks.

File: src/components/safe-messages/Msg/index.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { TextField } from '@mui/material'
+import type { ReactElement } from 'react'
+import type { SafeMessage } from '@safe-global/safe-gateway-typescript-sdk'
+
+import css from './styles.module.css'
+
+const Msg = ({ message }: { message: SafeMessage['message'] }): ReactElement => {
+  const isTextMessage = typeof message === 'string'
+
+  const readableData = isTextMessage ? message : JSON.stringify(message, null, 2)
```
Comment (katspaugh) [inline src/components/safe-messages/Msg/index.tsx:10]: Memoize perhaps?

File: src/components/safe-messages/MsgList/index.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import type { SafeMessageListPage } from '@safe-global/safe-gateway-typescript-sdk'
+
+import { TxListGrid } from '@/components/transactions/TxList'
+import MsgListItem from '@/components/safe-messages/MsgListItem'
+
+const MsgList = ({ items }: { items: SafeMessageListPage['results'] }): ReactElement => {
+  return (
+    <TxListGrid>
+      {items.map((item, i) => (
+        <MsgListItem item={item} key={i} />
```
Comment (katspaugh) [inline src/components/safe-messages/MsgList/index.tsx:11]: Do they have an id or something else to use for the key?

File: src/components/safe-messages/MsgModal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    return messageHash ?? generateSafeMessageHash(safe, decodedMessage)
+  }, [messageHash, safe, decodedMessage])
+
+  // Get already proposed message
+  const [alreadyProposedMessage] = useAsync<SafeMessage | Omit<SafeMessage, 'type'>>(() => {
+    const localMessage = messages.page?.results
+      .filter(isSafeMessageListItem)
+      .find((message) => message.messageHash === messageHash)
+
+    return localMessage ? Promise.resolve(localMessage) : getSafeMessage(safe.chainId, safeMessageHash)
```
Comment (katspaugh) [inline src/components/safe-messages/MsgModal/index.tsx:90]: In my tests, it was always trying to fetch the message and getting a 404 when I was signing a new message. Is this intended?

File: src/components/safe-messages/MsgModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  }, [safe.chainId, messageHash, safeMessageHash])
+
+  const hasSigned = !!alreadyProposedMessage?.confirmations.some(({ owner }) => owner.value === wallet?.address)
+
+  const isDisabled = isWrongChain || !isOwner || hasSigned || !web3
+
+  const onSign = async () => {
```
Comment (katspaugh) [inline src/components/safe-messages/MsgModal/index.tsx:97]: useCallback?

File: src/components/transactions/SignedMessagesHelpLink/index.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+  if (safeMessagesCount === 0) {
+    return null
+  }
+
+  return (
+    <Box display="flex" alignItems="center" gap={1}>
+      <SvgIcon component={InfoIcon} inheritViewBox color="border" fontSize="small" />
+      {/* TODO: Add link to help article */}
+      <ExternalLink noIcon href="#">
```
Comment (katspaugh) [inline src/components/transactions/SignedMessagesHelpLink/index.tsx:19]: Do we have the article yet?

File: src/hooks/useMsgBusTracking.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  [TxEvent.SIGNED]: WALLET_EVENTS.OFF_CHAIN_SIGNATURE,
+  [TxEvent.PROCESSING]: WALLET_EVENTS.ON_CHAIN_INTERACTION,
+}
+
+const safeMsgEvents = {
+  [SafeMsgEvent.PROPOSE]: MESSAGE_WALLET_EVENTS.SIGN_MESSAGE,
+  [SafeMsgEvent.CONFIRM_PROPOSE]: MESSAGE_WALLET_EVENTS.CONFIRM_MESSAGE,
+}
+
+export const useMsgBusTracking = (): void => {
```
Comment (katspaugh) [inline src/hooks/useMsgBusTracking.ts:16]: That's a strange name. If it's tracking both transactions and messages, `MsgBus` is wrong.

File: src/services/analytics/events/wallet.ts [hunk 8]
```diff
@@ -24,3 +24,16 @@ export const WALLET_EVENTS = {
     category: WALLET_CATEGORY,
   },
 }
+
+export const MESSAGE_WALLET_EVENTS = {
```
Comment (katspaugh) [inline src/services/analytics/events/wallet.ts:28]: Why a separate export?

File: src/components/safe-messages/MsgList/index.tsx (hunk 3 above)
Comment (iamacook) [reply src/components/safe-messages/MsgList/index.tsx:11]: The array is made up of messages and date labels. We could use something like `isSafeMessageListItem(item) ? item.messageHash : item.timestamp` but I don't think it's worth it.

File: src/components/transactions/SignedMessagesHelpLink/index.tsx (hunk 6 above)
Comment (iamacook) [reply src/components/transactions/SignedMessagesHelpLink/index.tsx:19]: Not yet. I've made @johannesmoormann aware and have a reminder.

File: src/hooks/useMsgBusTracking.ts (hunk 7 above)
Comment (iamacook) [reply src/hooks/useMsgBusTracking.ts]: I've reverted the changes and created a separate one specifically for messages.

File: src/services/analytics/events/wallet.ts (hunk 8 above)
Comment (iamacook) [reply src/services/analytics/events/wallet.ts]: I wanted to separate them. I've removed the separate export though.

File: src/components/safe-messages/MsgModal/index.tsx (hunk 4 above)
Comment (iamacook) [reply src/components/safe-messages/MsgModal/index.tsx:90]: Yes, we need to determine whether to propose or confirm the message. They are different endpoints.

**Review by katspaugh — APPROVED** [review]

File: docs/eip-1271.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+A common pitfall is that the `message`, `messageHash` and `safeMessageHash` gets mixed up.
+Currently when integrating a dApp, the `safeMessageHash` only needs to be used to fetch data from our transaction-service.
+In all other cases (verifying signatures, signing a message) the `messageHash` or the `message` is correct.
+
+## Restrictions
+
+- Only Safes with version `>1.0.0` are supported
+- Currently this is only implemented on staging services
+- Currently a patched / experimental version of the interface is required, which can be found [here](https://eip_1271--webcore.review-web-core.5afe.dev/)
+- Currently a patched / experimental version of the WalletConnect app is required, which can be found [here](https://eip_1271--webcore.review-web-core.5afe.dev/share/safe-app?appUrl=https%3A%2F%2Fpr608--safereactapps.review-react-hr.5afe.dev%2Fwallet-connect&chain=gor)
```
[the remark this answers is not in the document]
Comment (iamacook) [reply docs/eip-1271.md]: The two links have been removed as this will soon be merged but I will take this into account in the future.

**Review by katspaugh — APPROVED** [review]

**Pull request merged**